### ISSUE TITLE
In the UDT 4.11 library there was a use after free issue in the connect function.

### ISSUE THREAD

**Issue by whtghst1** [Haivision/srt#342]
use-after-free (in case of a connection time out, the failed socket representation was not removed from "waiting for connection" list, therefore it was accessed by the worker/garbage collection thread even after is was deleted)

The following addresses this.

```
if (CTimer::getTime() > ttl)
       {
+         // Stop waiting for this connection
+         m_pRcvQueue->removeConnector(m_SocketID);
+
          // timeout
          e = CUDTException(1, 1, 0);
          break;
}
```

**Comment by whtghst1** [comment]
I clone the source and investigate the garbage collect and see where it might also affect other areas in the `UDT::startConnect()` . I will the attach a proposed patch to the issue.

**Comment by ethouris** [comment]
Great, thanks! I will stay tuned.

**Comment by whtghst1** [comment]
This was fun since you have really refactored the connection code for SRT from UDT.

Issues where I see this problem for a blocking SRT socket during the UDT::startConnect:

1. When in the CUDT::ProcessConnectResponse and CUDT::processRendezvous calls and the connection status is returned CONN_REJECT the socket is only removed from the receive queue if the reject happens during those functions call to CUDT::postConnect and that calls subsequent call to CUDT::prepareConnectionObjects is false.  There are many time other time in those functions return CONN_REJECT and have not removed the socket id from the receive queue. This can cause the problem with the garbage collection cleanup with use after free.

2. When check check on ttl this breaking of the while loop does not remove the socket id from the received queue. This can cause the problem as well.

I now need to work through the socket event checks to see how non blocking sockets handle these conditions.

I have a patch that addresses the blocking socket connect, but need to ensure about the non-blocking before I put it here.

**Comment by whtghst1** [comment]
Look like the the async connect and response underneath uses the same process functions. So the patch should fix the problem for both scenarios.

The following patch from current master should address:

```
@@ -2811,10 +2811,14 @@ void CUDT::startConnect(const sockaddr* serv_addr, int32_t forced_isn)
             break;
         }
 
         if (CTimer::getTime() > ttl)
         {
+			//Remove the connector from the RcvQueue for TTL on blocking connect.
+			//All other breaks from this loop have already done this through CUDT::processConnectResponse
+			m_pRcvQueue->removeConnector(m_SocketID, true);
+
             // timeout
             e = CUDTException(MJ_SETUP, MN_TIMEOUT, 0);
             break;
         }
     }
@@ -2970,12 +2974,15 @@ EConnectStatus CUDT::processRendezvous(ref_t<CPacket> reqpkt, const CPacket& res
     cookieContest();
 
     // We know that the other side was contacted and the other side has sent
     // the handshake message - we know then both cookies. If it's a draw, it's
     // a very rare case of creating identical cookies.
-    if ( m_SrtHsSide == HSD_DRAW )
-        return CONN_REJECT;
+	if (m_SrtHsSide == HSD_DRAW)
+	{
+		m_pRcvQueue->removeConnector(m_SocketID, synchro);
+		return CONN_REJECT;
+	}
 
     UDTRequestType rsp_type = URQ_ERROR_INVALID; // just to track uninitialized errors
 
     // We can assume that the Handshake packet received here as 'response'
     // is already serialized in m_ConnRes. Check extra flags that are meaningful
@@ -2995,20 +3002,22 @@ EConnectStatus CUDT::processRendezvous(ref_t<CPacket> reqpkt, const CPacket& res
     m_ConnReq.m_extension = needs_extension;
 
     if (rsp_type > URQ_FAILURE_TYPES)
     {
         HLOGC(mglog.Debug, log << "processRendezvous: rejecting due to switch-state response: " << RequestTypeStr(rsp_type));
+		m_pRcvQueue->removeConnector(m_SocketID, synchro);
         return CONN_REJECT;
     }
 
     // This must be done before prepareConnectionObjects().
     applyResponseSettings();
 
     // This must be done before interpreting and creating HSv5 extensions.
     if ( !prepareConnectionObjects(m_ConnRes, m_SrtHsSide, 0))
     {
         HLOGC(mglog.Debug, log << "processRendezvous: rejecting due to problems in prepareConnectionObjects.");
+		m_pRcvQueue->removeConnector(m_SocketID, synchro);
         return CONN_REJECT;
     }
 
     // Case 2.
     if ( needs_hsrsp )
@@ -3018,10 +3027,11 @@ EConnectStatus CUDT::processRendezvous(ref_t<CPacket> reqpkt, const CPacket& res
         // This means that we have received HSREQ extension with the handshake, so we need to interpret
         // it and craft the response.
         if ( !interpretSrtHandshake(m_ConnRes, response, kmdata, &kmdatasize) )
         {
             HLOGC(mglog.Debug, log << "processRendezvous: rejecting due to problems in interpretSrtHandshake.");
+			m_pRcvQueue->removeConnector(m_SocketID, synchro);
             return CONN_REJECT;
         }
 
         // No matter the value of needs_extension, the extension is always needed
         // when HSREQ was interpreted (to store HSRSP extension).
@@ -3031,10 +3041,11 @@ EConnectStatus CUDT::processRendezvous(ref_t<CPacket> reqpkt, const CPacket& res
 
         rpkt.setLength(m_iMaxSRTPayloadSize);
         if (!createSrtHandshake(reqpkt, Ref(m_ConnReq), SRT_CMD_HSRSP, SRT_CMD_KMRSP, kmdata, kmdatasize))
         {
             HLOGC(mglog.Debug, log << "processRendezvous: rejecting due to problems in createSrtHandshake.");
+			m_pRcvQueue->removeConnector(m_SocketID, synchro);
             return CONN_REJECT;
         }
 
         // This means that it has received URQ_CONCLUSION with HSREQ, agent is then in RDV_FINE
         // state, it sends here URQ_CONCLUSION with HSREQ/KMREQ extensions and it awaits URQ_AGREEMENT.
@@ -3152,12 +3163,15 @@ EConnectStatus CUDT::processConnectResponse(const CPacket& response, CUDTExcepti
     // Returned values:
     // - CONN_REJECT: there was some error when processing the response, connection should be rejected
     // - CONN_ACCEPT: the handshake is done and finished correctly
     // - CONN_CONTINUE: the induction handshake has been processed correctly, and expects CONCLUSION handshake
 
-   if (!m_bConnecting)
-      return CONN_REJECT;
+	if (!m_bConnecting)
+	{
+		m_pRcvQueue->removeConnector(m_SocketID, synchro);
+		return CONN_REJECT;
+	}
 
    // This is required in HSv5 rendezvous, in which it should send the URQ_AGREEMENT message to
    // the peer, however switch to connected state. 
    HLOGC(mglog.Debug, log << "processConnectResponse: TYPE:" << MessageTypeStr(response.getType(), response.getExtendedType()));
    //ConnectStatus res = CONN_REJECT; // used later for status - must be declared here due to goto POST_CONNECT.
@@ -3206,31 +3220,35 @@ EConnectStatus CUDT::processConnectResponse(const CPacket& response, CUDTExcepti
 
    if ( !response.isControl(UMSG_HANDSHAKE) )
    {
        LOGC(mglog.Error, log << CONID() << "processConnectResponse: received non-addresed packet not UMSG_HANDSHAKE: "
            << MessageTypeStr(response.getType(), response.getExtendedType()));
+	   m_pRcvQueue->removeConnector(m_SocketID, synchro);
        return CONN_REJECT;
    }
 
    if ( m_ConnRes.load_from(response.m_pcData, response.getLength()) == -1 )
    {
        // Handshake data were too small to reach the Handshake structure. Reject.
        LOGC(mglog.Error, log << CONID() << "processConnectResponse: HANDSHAKE data buffer too small - possible blueboxing. Rejecting.");
+	   m_pRcvQueue->removeConnector(m_SocketID, synchro);
        return CONN_REJECT;
    }
 
    HLOGC(mglog.Debug, log << CONID() << "processConnectResponse: HS RECEIVED: " << m_ConnRes.show());
    if ( m_ConnRes.m_iReqType > URQ_FAILURE_TYPES )
    {
+	   m_pRcvQueue->removeConnector(m_SocketID, synchro);
        return CONN_REJECT;
    }
 
    if ( size_t(m_ConnRes.m_iMSS) > CPacket::ETH_MAX_MTU_SIZE )
    {
        // Yes, we do abort to prevent buffer overrun. Set your MSS correctly
        // and you'll avoid problems.
        LOGC(mglog.Fatal, log << "MSS size " << m_iMSS << "exceeds MTU size!");
+	   m_pRcvQueue->removeConnector(m_SocketID, synchro);
        return CONN_REJECT;
    }
 
    // (see createCrypter() call below)
    //
@@ -3241,10 +3259,11 @@ EConnectStatus CUDT::processConnectResponse(const CPacket& response, CUDTExcepti
    {
        // SANITY CHECK: A rendezvous socket should reject any caller requests (it's not a listener)
        if (m_ConnRes.m_iReqType == URQ_INDUCTION)
        {
            LOGC(mglog.Error, log << CONID() << "processConnectResponse: Rendezvous-point received INDUCTION handshake (expected WAVEAHAND). Rejecting.");
+		   m_pRcvQueue->removeConnector(m_SocketID, synchro);
            return CONN_REJECT;
        }
 
        // The procedure for version 5 is completely different and changes the states
        // differently, so the old code will still maintain HSv4 the old way.
```

**Comment by whtghst1** [comment]
BTW...I was only able to test build this patch on Windows platform. I do not have time to spin up another platform to test today.

**Comment by ethouris** [comment]
If you can't build for another platform, then try at least to compile on Cygwin :)
